### ISSUE TITLE
Strange overflow error?

### ISSUE THREAD

**Issue by gusmanb** [hsoft/collapseos#35]
I'm having a strange error when compiling with ZASM: Error 20 on line 206

From the source of ZASM I see 0x14 is overflow, but it does not make sense as the line where it's thrown is this:

    CP 0x3A

Also, I noticed that the error changed position when I corrected some .db entries (previously was on 212, after correcting the .db format it changed to 206 but I added no new line), so I think it there must be an error in other place.

The offending code is here: https://github.com/gusmanb/tinybasic/blob/master/TinyBasicCOLLAPSEOS.asm

Any idea what's happening?

**Comment by schierlm** [comment]
I am pretty sure that ZASM (especially its error handling) cannot cope with input files that are larger than 64KB, or longer than 256 lines.

Rationale: You cannot pack such large files into a CFS, and the increment instruction in line https://github.com/hsoft/collapseos/blob/master/apps/zasm/io.asm#L120 is an 8-bit increment.

I tried to remove all code after line 243 (but left labels), then removed most comment lines, and after a few replacements of constants in the form `0x0`xxxx, by removing the extra zero, it compiled for me. Of course it does not run any more.

Perhaps try to split it up in smaller files, and use includes in glue.asm to link them together again.
Not sure how soon you will hit the binary size limits. Perhaps you even need to assemble it separately and build your own jump table (similar to how userspace uses a jump table to jump into the kernel) to get it compiled within zasm's limitations.

**Comment by gusmanb** [comment]
Thanks for the info.

Are these limitations documented anywhere? The binary once compiled is 2Kb long so not sure if this is going to work...

Cheers.

**Comment by schierlm** [comment]
The only documentation I'm aware of is [in the source code](https://github.com/hsoft/collapseos/blob/83b314c450d276d61672c113c49631191ccc1c4b/apps/zasm/glue.asm#L49-L68) and I would assume that if you run against one of them, compilation will fail.

**Comment by hsoft** [comment]
This issue passed under my radar, but thanks @schierlm, you are correct about current limitations in zasm and CFS.

**Comment by schierlm** [comment]
The TinyBASIC binary is working.

Split up source: https://github.com/schierlm/CollapseOS-Web-Emulator/tree/master/basic

Web emulator to test it: https://schierlm.github.io/CollapseOS-Web-Emulator/

**Comment by schierlm** [comment]
But indeed there is something strange in either TinyBASIC or zasm. Last thing I changed before uploading my BASIC source in the previous comment was to remove a workaround I added to PRINT. Afterwards, PRINT still works fine (and I only tested that), but now FOR crashes... When adding `.db 0, 0, 0, 0` to the end of `print.asm`, in the resulting binary, FOR works again.

(Sure this does not have to be an issue in zasm, it could as well be that due to these 4 bytes, some other instruction gets aligned slightly differently in memory and triggers a bug elsewhere. Still strange to me.)
